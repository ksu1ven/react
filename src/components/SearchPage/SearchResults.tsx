import { Animal } from '../../utils/types';
import pawImg from '../../assets/paw.png';
import { useContext, useEffect } from 'react';
import { SetURLSearchParams, useNavigate } from 'react-router-dom';
import { updateQueryParams } from '../../utils/helpFunctions';
import { SearchResultsContext } from './Contexts';

interface Props {
  params: URLSearchParams;
  setParams: SetURLSearchParams;
}

function SearchResults(props: Props) {
  const { params, setParams } = props;
  const searchResultsArray = useContext(SearchResultsContext);
  const navigate = useNavigate();
  function checkDescription(animal: Animal) {
    const descriptionArr: string[] = [];
    Object.entries(animal).forEach((el) => {
      if (el[1] === true) {
        descriptionArr.push(el[0]);
      }
    });
    return descriptionArr.length
      ? descriptionArr.join(', ')
      : "API don't give us description :(";
  }

  useEffect(() => {
    navigate('?' + props.params, { replace: true });
  }, [navigate, props.params]);

  return (
    <div
      className="w-1/3 m-auto mb-10 flex flex-col gap-5 "
      data-testid="search-results"
    >
      {searchResultsArray.length ? (
        searchResultsArray.map((el) => {
          const animalDescription = checkDescription(el);
          return (
            <div
              key={el.uid}
              className="flex justify-between gap-5 cursor-pointer"
              data-testid="card"
              onClick={() => {
                setParams(updateQueryParams(params, 'details', el.name));
              }}
            >
              <img src={pawImg} alt="animal picture" className="w-16 h-16" />
              <div className="grow">
                <h2 className="font-extrabold mb-1">{el.name}</h2>
                <p>Description: {animalDescription}</p>
              </div>
            </div>
          );
        })
      ) : (
        <p className="text-3xl" data-testid="no-found">
          Nothing found:(
        </p>
      )}
    </div>
  );
}

export default SearchResults;
