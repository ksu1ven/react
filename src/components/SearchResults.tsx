import React from 'react';
import { useRouter } from 'next/router';
import { updateQueryParams } from '../utils/helpFunctions';
import { Animal } from '../utils/types';
import Image from 'next/image';
import pawImg from '../assets/paw.png';

function SearchResults(props: Record<'searchResults', readonly Animal[]>) {
  const { searchResults } = props;
  const router = useRouter();

  function checkDescription(animal: Animal) {
    const descriptionArr: string[] = [];
    Object.entries(animal).forEach((el) => {
      if (el[1] === true) {
        descriptionArr.push(el[0]);
      }
    });
    return descriptionArr.length
      ? descriptionArr.join(', ')
      : "API don't give us description :(";
  }

  return (
    <div
      className="w-1/3 m-auto mb-10 flex flex-col gap-5 "
      data-testid="search-results"
    >
      {searchResults.length ? (
        searchResults.map((el) => {
          const animalDescription = checkDescription(el);
          return (
            <div
              key={el.uid}
              className="flex justify-between gap-5 cursor-pointer"
              data-testid="card"
              onClick={() => {
                router.push(
                  '?' + updateQueryParams(router.query, 'details', el.name)
                );
              }}
            >
              <Image
                src={pawImg}
                alt="animal picture"
                className="w-16 h-16"
                width={100}
                height={100}
              />
              <div className="grow">
                <h2 className="font-extrabold mb-1">{el.name}</h2>
                <p>Description: {animalDescription}</p>
              </div>
            </div>
          );
        })
      ) : (
        <p className="text-3xl" data-testid="no-found">
          Nothing found:(
        </p>
      )}
    </div>
  );
}

export default SearchResults;
