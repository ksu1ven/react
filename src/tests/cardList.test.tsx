import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import SearchResults from '../components/SearchResults';
import { createCardsListResponseMock } from './_mocks';

vi.mock('next/router', () => vi.importActual('next-router-mock'));

describe('Tests for the Card List component', () => {
  it('Component renders the specified number of cards - 7', async () => {
    const searchResults = render(
      <SearchResults
        searchResults={createCardsListResponseMock(7, 10, false)}
      />
    );
    expect(searchResults.baseElement).toMatchSnapshot();
    expect(await screen.findAllByText(/testCard/)).toHaveLength(7);
  });

  it('An appropriate message is displayed if no cards are present', async () => {
    const searchResults = render(<SearchResults searchResults={[]} />);

    expect(searchResults.baseElement).toMatchSnapshot();
    expect(screen.queryByText('Nothing foundddddddddddd:(')).toBeFalsy();
    expect(screen.getByText('Nothing found:(')).toBeTruthy();
    expect(screen.queryByText(/testCard/)).toBeFalsy();
  });
});
