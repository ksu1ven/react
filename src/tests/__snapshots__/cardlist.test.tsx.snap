// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Card List component > An appropriate message is displayed if no cards are present 1`] = `
<body>
  <div>
    <div
      class="w-1/3 m-auto mb-10 flex flex-col gap-5 "
      data-testid="search-results"
    >
      <p
        class="text-3xl"
        data-testid="no-found"
      >
        Nothing found:(
      </p>
    </div>
  </div>
</body>
`;

exports[`Tests for the Card List component > Component renders the specified number of cards - 7 1`] = `
<body>
  <div>
    <div
      class="w-1/3 m-auto mb-10 flex flex-col gap-5 "
      data-testid="search-results"
    >
      <div
        class="flex justify-between gap-5 cursor-pointer"
        data-testid="card"
      >
        <img
          alt="animal picture"
          class="w-16 h-16"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75"
          srcset="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=128&q=75 1x, /_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75 2x"
          style="color: transparent;"
          width="100"
        />
        <div
          class="grow"
        >
          <h2
            class="font-extrabold mb-1"
          >
            testCard-0
          </h2>
          <p>
            Description: 
            avian
          </p>
        </div>
      </div>
      <div
        class="flex justify-between gap-5 cursor-pointer"
        data-testid="card"
      >
        <img
          alt="animal picture"
          class="w-16 h-16"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75"
          srcset="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=128&q=75 1x, /_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75 2x"
          style="color: transparent;"
          width="100"
        />
        <div
          class="grow"
        >
          <h2
            class="font-extrabold mb-1"
          >
            testCard-1
          </h2>
          <p>
            Description: 
            avian
          </p>
        </div>
      </div>
      <div
        class="flex justify-between gap-5 cursor-pointer"
        data-testid="card"
      >
        <img
          alt="animal picture"
          class="w-16 h-16"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75"
          srcset="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=128&q=75 1x, /_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75 2x"
          style="color: transparent;"
          width="100"
        />
        <div
          class="grow"
        >
          <h2
            class="font-extrabold mb-1"
          >
            testCard-2
          </h2>
          <p>
            Description: 
            avian
          </p>
        </div>
      </div>
      <div
        class="flex justify-between gap-5 cursor-pointer"
        data-testid="card"
      >
        <img
          alt="animal picture"
          class="w-16 h-16"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75"
          srcset="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=128&q=75 1x, /_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75 2x"
          style="color: transparent;"
          width="100"
        />
        <div
          class="grow"
        >
          <h2
            class="font-extrabold mb-1"
          >
            testCard-3
          </h2>
          <p>
            Description: 
            avian
          </p>
        </div>
      </div>
      <div
        class="flex justify-between gap-5 cursor-pointer"
        data-testid="card"
      >
        <img
          alt="animal picture"
          class="w-16 h-16"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75"
          srcset="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=128&q=75 1x, /_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75 2x"
          style="color: transparent;"
          width="100"
        />
        <div
          class="grow"
        >
          <h2
            class="font-extrabold mb-1"
          >
            testCard-4
          </h2>
          <p>
            Description: 
            avian
          </p>
        </div>
      </div>
      <div
        class="flex justify-between gap-5 cursor-pointer"
        data-testid="card"
      >
        <img
          alt="animal picture"
          class="w-16 h-16"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75"
          srcset="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=128&q=75 1x, /_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75 2x"
          style="color: transparent;"
          width="100"
        />
        <div
          class="grow"
        >
          <h2
            class="font-extrabold mb-1"
          >
            testCard-5
          </h2>
          <p>
            Description: 
            avian
          </p>
        </div>
      </div>
      <div
        class="flex justify-between gap-5 cursor-pointer"
        data-testid="card"
      >
        <img
          alt="animal picture"
          class="w-16 h-16"
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75"
          srcset="/_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=128&q=75 1x, /_next/image?url=%2Fsrc%2Fassets%2Fpaw.png&w=256&q=75 2x"
          style="color: transparent;"
          width="100"
        />
        <div
          class="grow"
        >
          <h2
            class="font-extrabold mb-1"
          >
            testCard-6
          </h2>
          <p>
            Description: 
            avian
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
